name_tracker.go

```diff
@@ -6,14 +6,47 @@ package quamina
 // Quamina's MatchesForEvent API is strongly related to the number of fields it has to look at.  Therefore, a
 // Flattener should call NameTracker for each field it encounters and if it returns false, skip that field
 // and not add it to the flattened output. Here's a clarifying example; consider this JSON event:
-//  {"a": {"b": 1, "c": 2}}
+//
+//	{"a": {"b": 1, "c": 2}}
+//
 // Assuming NameTracker always returns positive, the flattened output should look like
-//  "a\nb" 1
-//  "a\nc" 2
+//
+//	"a\nb" 1
+//	"a\nc" 2
+//
 // Note that the values stored in NameTracker are not the full paths "a\nb" and "a\nc", but the path segments
 // "a", "b", and "c".  This reduces the work the Flattener has to do - whenever it processes an object, it need
 // only look up the member names in the NameTracker. A little thought reveals that this can produce false
 // positives and some potential wasted work, but is a good trade-off.
 type NameTracker interface {
 	IsNameUsed(label []byte) bool
+
+	// This should be actually part of matcher interface and not NameTracker.
+	GetHierarchalNameTracker() HierarchalNameTracker
+}
+
+type HierarchalNameTracker interface {
+	// Get returns another level of the hierarchy, referred as "Node"
+	// If a node is returned we will need to traverse into (in JSON/CBOR/ProtoBuf/etc..)
+	Get(member []byte) (HierarchalNameTracker, bool)
+
+	// IsRoot - are we root node?
+	// NOTE: need for early exit, can be solved differently maybe.
+	IsRoot() bool
+
+	// Is the current property needs to be plucked as a Quamina's Field.
```

Comment:
Not sure what "property" means… field? Probably "processed" rather than "plucked"?

Reply:
It should be actually (after the learnings today on Quamin's internal terminology) -
```
// Do we need to select the given segment from event as Quamina's Field.
IsSegmentUsed(segment []byte) bool
```

Reply:
renamed